### PR TITLE
20190331 NUS templates

### PR DESCRIPTION
**Pull request by ccewuz**

### CODE REVIEW

**Comment by rjchow** [comment]
do make sure linting passes

**Comment by rjchow** [comment]
is this PR ready for review? please self-assess against this checklist that i've written up recently first - thanks!

**Comment by ccewuz** [comment]
[NUS Restricted]

Yes, ready for review. Some questions on self assessment:

Certificate Template

  *   [ ] Sample certificate file included for each template – Q: where should I put the sample files?
  *   [ ] Sample certificates must obviously be a sample certificate
     *   [ ] Obviously fictitious name
     *   [ ] Obviously sample signatory images
  *   [ ] No fixed-size raster images as part of certificate layout
  *   [ ] Mobile responsive design
  *   [ ] Date parsing should be localised to template author’s timezone
  *   [ ] Webpack chunking code is correct
     *   [ ] Has chunking code
     *   [ ] Same chunking code as the other certificates belonging to that institute
  *   [ ] Certificate Store Addresses have been updated
     *   [ ] Template Whitelist
     *   [ ] Registry
  *   [ ] Template should not be using resources(images etc.) on the website outside of their own folder (e.g institute logo shouldn’t be used from /static because there’s no guarantee it will not change)
Pre-merge checks

  *   [ ] Ensure that your code has been rebased<https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request> on top of latest OpenCerts master
  *   [ ] Linter issues resolved (Run npm run lint:fix to see issues)
  *   [ ] npm run test passes – Q: Failed but it seems not because of my code. Please refer to npmtest.log
  *   [ ] npm run test:integration passes – Q: Failed with error: missing script: test:integration
  *   [ ] Travis Build passes<https://docs.travis-ci.com/user/for-beginners/>


Thanks & Regards,
Zehe

**Comment by rjchow** [comment]
Sorry I missed out one point in the checklist - have updated the checklist now

Each template should have an [integration test](https://docs.opencerts.io/integration_test.html) located in its directory, and that test should use the sample certificate located in the directory to run some rendering tests - these tests are to protect your institutions' interests and should be thorough in order to ensure that their rendering is not broken by any code changes unintentionally.

Please refer to [docs](https://docs.opencerts.io/integration_test.html) for more information on these tests

**Comment by ccewuz** [comment]
[NUS Restricted]

Thanks. Added test scripts and sample files. Integration test passed.

Thanks & Regards,
Zehe

File: src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      type = "";
+      break;
+  }
+  const html = (
+    <img
+      style={{
+        width: imageData.width + imageData.unit.toLowerCase(),
+        height: imageData.height + imageData.unit.toLowerCase()
+      }}
+      src={`data:image/${type};base64,${imageData.data}`}
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js]: You can store the entire string in the cret file so you dont have to break it up and join it back later. See the image in NP sample cert: https://github.com/OpenCerts/opencerts-website/blob/master/src/components/CertificateTemplates/tlds/sg/edu/np/NP-AA2018-MAIN/NP_Certs_FT_MAIN_2018.opencert

File: src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import React from "react";
+
+export const isoDateToLocal = isoDate =>
+  `${isoDate.slice(8, 10)}/${isoDate.slice(5, 7)}/${isoDate.slice(0, 4)}`;
+
+export const isoDateToLocalLong = isoDate => {
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js]: Use the library `moment-timezone`, see it at https://github.com/OpenCerts/opencerts-website/blob/master/src/components/CertificateTemplates/tlds/sg/edu/np/common/optioncertificate.js

File: src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
+import React from "react";
+
+export const isoDateToLocal = isoDate =>
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js]: See below comment

File: src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    "August",
+    "September",
+    "October",
+    "November",
+    "December"
+  ][parseInt(isoDate.slice(5, 7), 10) - 1];
+  return `${day} ${month} ${year}`;
+};
+
+export const isBrowser = browser => {
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js]: Why do you need these? Possible to remove all the browser specific checks? Don't see it used anywhere.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.scss [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        div {
+            border: 1px solid blue; 
+        }
+    }
+    
+    font-family: Arial; 
+    
+    .a4-portrait { 
+        background: white; 
+        height: 29.7cm;
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.scss]: same problem with `cm`. Please make this mobile responsive.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+              {this.renderLogo()}
+              {this.renderVoid("0.93cm")}
+              {this.renderContent()}
+            </div>
+            <div style={{ border: "0px solid" }}>{this.renderSigs()}</div>
+          </article>
+        </div>
+      </div>
+    );
+    return html;
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]: No need for [].

```
return (<div className={cls("nus-degree")}>
        <div className={cls("a4-portrait")}>
          <article>
            <div style={{ height: "23.4cm", border: "0px solid" }}>
              {this.renderVoid("2.13cm")}
              {this.renderTitle()}
              {this.renderVoid("0.59cm")}
              {this.renderLogo()}
              {this.renderVoid("0.93cm")}
              {this.renderContent()}
            </div>
            <div style={{ border: "0px solid" }}>{this.renderSigs()}</div>
          </article>
        </div>
      </div>)
```

Use alternate way to render the spaces between.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+const cls = names => sassClassNames(names, scss);
+
+class Degree extends Component {
+  constructor(props) {
+    super(props);
+    this.dataSource = this.props.dataSource;
+  }
+
+  // render an empty div with specified height
+  renderVoid = height => (
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]: Use paddings/classes instead.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    return html;
+  }
+
+  // render signatures
+  renderSigs = () => {
+    let sig1;
+    let sig2;
+    if (this.dataSource.additionalData.images) {
+      sig1 = renderImage(this.dataSource.additionalData.images.TRUSTEES);
+      sig2 = renderImage(this.dataSource.additionalData.images.PRESIDENT);
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]: No need for renderImage, see comment in that. Simply write the html in where you need the image.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+              </div>
+            </td>
+          </tr>
+          <tr>
+            <td align="center">
+              {" "}
+              {/* signature 2 */}
+              <div className={cls("cert-sig")}>
+                <br />
+                {sig2}
```
Comment (yehjxraymond) [inline src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]:
```
<img src="{imagesource}" />
```

will do the trick

**Comment by ccewuz** [comment]
[NUS Restricted]

Are those changes required or recommended?

Thanks & Regards,
Zehe

**Comment by yehjxraymond** [comment]
Required.

**Comment by ccewuz** [comment]
[NUS Restricted]

Questions and comments:


1)     dataSource is not required for some of methods. So will not pass them to all methods, only one is applicable which I will update

2)     Any reason why array is not allowed in rendering? Isn’t it valid?

3)     CM is used in CSS because NUS requirement is we want e-certificates look as much like paper copy as possible. Don’t think grid system can do that. I will double check with business owner whether they insist on this.

4)     Regular spaces can’t work the same as &nbsp; as I have tested.

5)     Want to retain renderImage() as a routine to be re-used by all templates

Thanks & Regards,
Zehe

**Comment by yehjxraymond** [comment]
1)     dataSource is not required for some of methods. So will not pass them to all methods, only one is applicable which I will update

Sure

2)     Any reason why array is not allowed in rendering? Isn’t it valid?

It's used to map array of similar objects, as such they require a key.

3)     CM is used in CSS because NUS requirement is we want e-certificates look as much like paper copy as possible. Don’t think grid system can do that. I will double check with business owner whether they insist on this.

Look at your cert on mobile. Show that to your user.

4)     Regular spaces can’t work the same as &nbsp; as I have tested.

Spaces are fine. See the other templates. 

5)     Want to retain renderImage() as a routine to be re-used by all templates

You do not even need it. Found you a reference, look at it.

And there's no need to click on reply, it makes this page really long and unreadable.

**Comment by rjchow** [comment]
do let us know when you would like this to be reviewed again

**Comment by ccewuz** [comment]
You can leave it for now as we are doing enhancement. Anyway, we will stick to current design which is not “mobile responsive” as this is NUS’ requirement as confirmed by business owner.

Thanks & Regards,
Zehe


________________________________

Important: This email is confidential and may be privileged. If you are not the intended recipient, please delete it and notify us immediately; you should not copy or use it for any purpose, nor disclose its contents to any other person. Thank you.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+  // render degree and honours
+  renderDegree = degreeData => {
+    const html = [];
+    html.push(
+      <div className={cls("cert-degree")}>
+        {degreeData.degreeTitle.toUpperCase()}
+      </div>
+    );
+    let honorsTitle = degreeData.honours ? degreeData.honours : "";
+    if (honorsTitle) {
```
[the remark this answers is not in the document]
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]: resolved

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js (hunk 7 above)
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]: this uses style

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js (hunk 8 above)
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]: necessary as rendering image is used everywhere. Function is used reduce code, as multiple attributes are used in rendering.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js (hunk 9 above)
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]: same as above.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+              {" "}
+              {/* This is to certify that */}
+              <div className={cls("cert-content")} style={style1}>
+                This&nbsp;&nbsp;&nbsp;is&nbsp;&nbsp;&nbsp;to&nbsp;&nbsp;&nbsp;certify&nbsp;&nbsp;&nbsp;that
+              </div>
+            </td>
+          </tr>
+          <tr>
+            <td>
+              {" "}
```
[the remark this answers is not in the document]
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/NUS-K1-2019/degree.js]: already removed.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js (hunk 1 above)
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js]: already done.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js (hunk 2 above)
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js]: done

File: src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js (hunk 4 above)
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js]: not useful for now. just stand by.

File: src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js (hunk 3 above)
Comment (ccewuz) [reply src/components/CertificateTemplates/tlds/sg/edu/nus/common/routines.js]: done

**Comment by ccewuz** [comment]
[NUS Restricted]

Hi, I have update the pull request with 10+ templates, for your review.

No idea why the following has been ongoing for days. Can you please look into it?

continuous-integration/travis-ci Expected — Waiting for status to be reported


Thanks & Regards,
Zehe

From: Wu Zehe
Sent: Tuesday, 23 April 2019 8:41 AM
To: OpenCerts/opencerts-website <reply@reply.github.com>; OpenCerts/opencerts-website <opencerts-website@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [OpenCerts/opencerts-website] 20190331 NUS templates (#135)

You can leave it for now as we are doing enhancement. Anyway, we will stick to current design which is not “mobile responsive” as this is NUS’ requirement as confirmed by business owner.

Thanks & Regards,
Zehe



________________________________

Important: This email is confidential and may be privileged. If you are not the intended recipient, please delete it and notify us immediately; you should not copy or use it for any purpose, nor disclose its contents to any other person. Thank you.

**Comment by yehjxraymond** [comment]
Your test are still failing: https://travis-ci.org/OpenCerts/opencerts-website/jobs/545054191
You still have merge conflict. 

Please rebase all your changes on top of the current master. Your code is from very long ago and it will likely have many issues merging. If need be, create separate PR with maximum of 5 templates or 25 changed files.

**Comment by ccewuz** [comment]
OK. Now all checks pass.

**Comment by yehjxraymond** [comment]
Refactoring them helps you figure out what should be the same and what shouldn't. Anyway the figure its not far from the quality gate, wont take long for you to fix that.

**Comment by ccewuz** [comment]
Managed to reduce duplication. But now stuck at "continuous-integration/travis-ci" again. I can't see details of the check. Can you help me check? Thanks.

**Comment by ccewuz** [comment]
OK now. All checks passed.

**Pull request closed without merging**